Make `dynamicsymbols._t` public and `dynamicsymbols` independent of `.t`

sidhu1012 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #17714

#### Brief description of what is fixed or changed

- The private attribute `dynamicsymbols._t` is made public and now it is `dynamicsymbols.t`. 
- To change the time variable in dynamicfunctions, a new parameter `time` is added to function parameters, changing `time` in dynamicsymbols won't effect other backend time variables
- To change internal time variable(used in various functions at backend) update `dynamicsymbols.t`.

#### Other comments
dynamicsymbols._t is still backwards compatible but can only be used as a symbol. Will be removed in future .

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.vector
    * The private attribute `dynamicsymbols._t` is made public and now it is `dynamicsymbols.t`. 
    *  To change the time variable in dynamicfunctions, a new parameter `time` is added to function parameters, changing `time` in dynamicsymbols won't effect other backend time variables
<!-- END RELEASE NOTES -->

moorepants committed:
Created a public constant for time in physics.vector

This introduces a constant for the time symbol used in physics.vector
and physics.mechanics and starts the deprecation of storing this value
on the dynamicsymbols function.

- TIME is defined in physics.vector.functions.TIME
- All uses of dynamicsymbols._t are removed, but this is still
  functional for backwards compatibility.
- Documentation showing the use two private attributes of dynamicsymbols
  is removed so we don't recommend it further.
- Documentation and a test is added showing how to set the time symbol
  to something else.

moorepants committed:
Included imports and redefined TIME.

moorepants committed:
Make sure setting TIME doesn't pollute other doctests.

moorepants committed:
Added warning about monkey patching TIME.

moorepants committed:
Improved docstring of dynamicsymbols.

moorepants committed:
Fixed doctest errors.

moorepants committed:
Merge branch 'master' into make-time-public

moorepants committed:
Added test for the old way to set time.

sidhu1012 committed:
Update rst, init of vector and fix dynamicsymbols._t

sidhu1012 committed:
update t in dynamicsymbols

sidhu1012 committed:
make t = dynamicsymbols._t in dynamicsymbols

sidhu1012 committed:
Merge branch 'master' into pr-17714

sidhu1012 committed:
fix code quality

sidhu1012 committed:
Remove unused imports

sidhu1012 committed:
Make `._t` unchangeable

sidhu1012 committed:
Change all instances of dynamicsymbols._t to TIME

sidhu1012 committed:
Remove trailing white space

sidhu1012 committed:
Remove unused imports

sidhu1012 committed:
Update test to check monkey patching

sidhu1012 committed:
update remaining instance of dynamicsymbols._t to TIME

sidhu1012 committed:
Update tests and comments to make thing clearer

sidhu1012 commented:
@moorepants It's up for review.

asmeurer commented on doc/src/modules/physics/vector/advanced.rst:
```diff
@@ -8,6 +8,48 @@ the features that will be implemented in the future will also be covered, along
+   >>> q = dynamicsymbols('q')
+   >>> q
+   q(t)
+   >>> list(q.free_symbols)[0] is TIME
```
I don't think TIME should be treated as singletonized. It's just defined as `Symbol('t')`. Another `Symbol('t')` should also work. And in general, singletonization isn't necessary for SymPy objects, because they are immutable, so regular equality comparison works fine.

asmeurer commented on doc/src/modules/physics/vector/advanced.rst:
```diff
@@ -8,6 +8,48 @@ the features that will be implemented in the future will also be covered, along
+time to something other than the default. It is possible to do so, but it
+requires monkey patching the variable in the module. **Note that this is
+fragile and you can likely encounter errors if the imports are not carefully
+done.** You can set a different symbol for time like so::
```
Just looking at the code below, this isn't going to work. If you want it to work, you have to replace 

```py
from sympy.physics.vector.functions import TIME

...

TIME
```

with 

```py
from sympy.physics.vector import functions

...

functions.TIME
```

Otherwise `TIME` will still be what it was imported as. If you want to support this, you need to make this basically the *only* supported way of accessing this API. It needs to work like that in non-library code as well, if that code is expected to support changing the time variable. 

Personally, I would recommend against supporting this. A much simpler way to get the same functionality would be to use a custom object for TIME and allow modifying the printer for it.

asmeurer replied:
Ironically, the old method, `dynamicsymbols._t`, actually does support this, because `_t` is used as an attribute. At best, if you want this to work this way, you should make it an attribute of an object rather than a top level name, to make it impossible to import it in a way that breaks monkeypatching (it would still break if someone did `t = dynamicsymbols.TIME`, but there's no way to avoid that while keeping with the immutable design, which is why I highly recommend doing the printer approach instead).

moorepants replied:
> A much more robust way to get the same functionality would be to use a custom object for TIME and allow modifying the printer for it.

This is worth some thought. Thanks.

asmeurer commented:
Can you reference the other issues/PRs in the pull request description? I know this has been discussed before.

sidhu1012 commented:
> Can you reference the other issues/PRs in the pull request description? I know this has been discussed before.

#17714 and #9818

moorepants commented:
Thanks for the tips @asmeurer, we'll see about adjusting the approach here.

sidhu1012 commented:
@moorepants @asmeurer If we create an object for `TIME` then we may be able to change `dynamicsymbols` time variable but 

```
>>> from sympy.physics.vector import dynamicsymbols, TIME #Time being a custom object
>>> v = dynamicsymbols('v')
>>> v.diff(TIME) # This won't work

```

moorepants commented:
The custom object for time could be a subclass of Symbol where the symbol name defaults to 't', I guess. The use case for being able to change the symbol for time are told fold: 1) you want something other than 't' to be printed and 2) you need two work with more than one unique time variable in the same script. I thought we supported changing the variable in the past for both reasons. But there isn't really any explicit use case driving either of these reasons.

The issue that this PR tries to solve is regarding the issue that the user needs access to 't' often in their scripts.

One solution is that if you need the variable for time available in the script you always have to do this:

```python
t = Symbol('t')
q = dynamicsymbols('q', time=t)
q.diff(t)
```

That's not so bad. And if the user doesn't need `t` expliclity in their script they don't need to set the kwarg in dyamicsymbols. I supposed we could just make `dynamicsymbols.t` instead of `dynamicsymbols._t`.

In a script with two time symbols we'd see:

```python
t = Symbol('t')
tau = Symbol('tau')
q_t = dynamicsymbols('q', time=t)
q_tau = dynamicsymbols('q', time=tau)
```

But the issue with this code is that any following methods that are called may use `dynamicsymbols._t` internally. So if you have:

```python
A = ReferenceFrame('A')
v = q_t*A.x + q_tau*A.y
v.dt(A)
```
That is ill defined because `dt()=.diff(dynamicsymbols._t, A)`.

We really need a use case for having more than one unique time variable in play before trying to support that. It may be that there is no reason to support it.

If all the user wants to do is make `t` print as `tau` then, as Aaron said, we can show them how to adjust printing.

sidhu1012 commented:
Yes, it might be a better solution.

sidhu1012 committed:
make dynamicsymbols.t public and make dynamicsymbols independent of dynamicsymbols.t

sidhu1012 commented:
@moorepants @asmeurer Review please! This way user can defined dynamic symbol with required symbol for time. If user want's to change the backend symbol for time he can do so by changing `dynamicsymbols.t` .

I'll fix failing tests tomorrow.

sidhu1012 committed:
Remove `real=True` from time and dynamicsymbols

sidhu1012 committed:
fix flake

sidhu1012 committed:
Fix advane rst

sidhu1012 committed:
Add `dynamicsymbols._t` back for backwards compatibility

sidhu1012 committed:
Add blank line in end

sidhu1012 commented:
@moorepants Done with changes! This way seems good to me, please review.

sidhu1012 commented:
@moorepants It's up for review.

sidhu1012 commented:
@moorepants Please review this PR.

moorepants commented:
I'm now doubting whether we should add the `time` kwarg. The prior behavior was:

```pycon
>>> dynamicsymbols._t = Symbol('tau')
>>> f = dynamicsymbols('f')
>>> f
f(tau)
>>> mechanics_function_that_uses_time_internally(f)
# some function of tau is returned
```

Setting `_t` effectively sets this value globally, right?

If we change `._t` to `.t` then it makes this attribute "public" which is a good thing. I'm fine with that.

But if we add the `time=` kwarg, there there are two ways to set this value and it also isn't clear if it sets it globally or not. Is this the behavior:

```pycon
>>> f = dynamicsymbols('f', time=Symbol('tau'))
>>> f
f(tau)
>>> g = dynamicsymbols('g')
>>> g
g(t)
```

Will `g` be a function of `tau` or `t`? And will functions that call `dynamicsymbols.t` internally get `tau` or `t`?

If we thave `time=` in `dynamicsymbols, then I think we need things like:

```
Vector.diff(time=Symbol('tau'))
KanesMethod(time=Symbol('tau'))
```

That would make things consistent.

But I don't think we need to two ways to do this. I think the simplest thing is to just change `._t` to `.t` and not add `time=`. It opens up too many confusions and issues with both.

sidhu1012 commented:
@moorepants Most of the users used `dynamicsymbols._t` to change the variable of time in `dynamicsymbols` but updating `dynamicsymbols._t` also updated all mechanics function that used time function internally, which in most cases isn't desired so I changed the functionality by introducing `time` kwarg which updates the symbols of time in `dynamicsymbols` and **makes dynamicsymbols independent of `dynamicsymbols._t`**(now public `dynamicsymbols.t`). 

The new functionality is

```
>>> f = dynamicsymbols('f', time=Symbol('tau'))
>>> f
f(tau)
>>> g = dynamicsymbols('g')         #Previous change of time wont effect 
>>> g
g(t)
```

Now if we update `dynamicsymbols.t`

```
>>> g = dynamicsymbols('g')
>>> g
g(t)
>>> dynamicsymbols.t = Symbol('tau')
>>> f = dynamicsymbols('f')
>>> f
f(t)         # dynamicsymbols.t have no effect on dynamicsymbols
```

Updating `dynamicsymbols.t` won't effect `dynamicsymbols` but would change all internal functions using time and similarly `dynamicsymbols time kwarg` won't effect `dynamicsymbols.t` or any internal function.

```
>>> dynamicsymbols.t = Symbol('tau')
>>> f = dynamicsymbols('f')
>>> f
f(t)
>>> mechanics_function_that_uses_time_internally(f)
# some function of tau is returned
```

moorepants commented:
I think the only change we should do is add `.t` which mirrors `._t`. The `time` kwarg is only useful if we redesign all of the relevant pieces of vector/mechanics to accept a `time` kwarg.

sidhu1012 commented:
> I think the only change we should do is add `.t` which mirrors `._t`. The `time` kwarg is only useful if we redesign all of the relevant pieces of vector/mechanics to accept a `time` kwarg.

@moorepants But I believe along with `.t` we should have `time` kwarg as mostly people use `dynamicsymbols._t` to change the time variable of dynamicsymbols, so I think changing `dynamicsymbols.t` just for that could mess up code for user without his knowing.

Also, only advance users would change time variable in `mechanics_function_that_uses_time_internally` for which they have `dynamicsymbols.t` but for regular users they could simply use time kwarg as they should not be concerned with `dynamicsymbols.t`.

moorepants commented:
> so I think changing dynamicsymbols.t just for that could mess up code for user without his knowing

What's an example where this would mess up code?

sidhu1012 commented:
> What's an example where this would mess up code?
By mess up code I don't mean it would definitely give something vague but I mean it may make things confusing for a simple user.

General user uses `._t` just to access time variable `t` not to update it in backend as seen in all code snippets, the most general use case of updating `._t` would be to change the time variable in dynamicsymbols , not in other functions using `._t` in backend but unknowingly the user may change `._t` in all functions using `._t` for such a small use case. 

For example
```
>>> m = dynamicsymbols('m')
#Now user wants to make dynamicsymbols a function of tau rather than t
>>> dynamicsymbols._t = Symbol('tau')
>>> n = dynamicsymbols('n')
>>>n
n(tau)
#But a general user here may not know(mostly won't know) that he just made all functions using `._t` use tau instead of t
>>> mechanics_function_that_uses_time_internally(n)
#some function of tau is returned instead of t
```

What I mean is, an advance user may know and understand the complication and serious effects of `dynamicsymbols._t` but a general user may just mess up his code for a small use case of changing time variable in dynamicsymbols.

So I suggest this may be better
```
>>> m = dynamicsymbols('m')
>>> n = dynamicsymbols('n', time=Symbol('tau'))  # Doesn't change any internal function using `._t`
>>> m
m(t)
>>> n 
n(tau)
>>> mechanics_function_that_uses_time_internally(n)
# Some function of t is returned as expected
```

sidhu1012 commented:
@moorepants What's the final take on it?
